### PR TITLE
Split files_sharing middelware

### LINKED ISSUE #19477: Federated Cloud Sharing without Resharing
A user reported:

> We are trying to set up an owncloud installation for guest users. They should not have fully privileged accounts, but should be only allowed to receive shares via federated cloud sharing. Resharing via links or internal sharing shall not be allowed. Now I figured out that the users are only able to receive federated shared, if the flag "allow users, to share contents via link" (the second shown in the screenshot) is set. Otherwise, the notification that they received a new share does not appear. But this also allows the users to share links what we would like to deny.

> For me this seems like a bug. Is there a way to fix this?

ihdfgbca

### Server configuration

**ownCloud version:** (see ownCloud admin page)
8.1.3

**Updated from an older ownCloud or fresh install:**
fresh install

@MorrisJobke 

00003923

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #19477

Since for external shares there is no need for link shares to be enabled
we should check which controller is actually being called.

This makes sure that in all cases we verify that the files_sharing app
is enabled. But only for the share controller (public shares) we check
if the API is enabled and if links are enabled.

TODO: 
- [x] add checks for federated sharing as well

### CODE REVIEW

File: apps/files_sharing/lib/middleware/sharingcheckmiddleware.php [hunk 1]
```diff
@@ -68,6 +68,13 @@ public function beforeController($controller, $methodName) {
 		if(!$this->isSharingEnabled()) {
 			throw new NotFoundException('Sharing is disabled.');
 		}
+
+		if ($controller instanceof \OCA\Files_Sharing\Controllers\ExternalSharesController) {
+			//TODO: Proper checks
+		} else if ($controller instanceof \OCA\Files_Sharing\Controllers\ShareController &&
+			$this->isLinkSharingEnabled()) {
+			throw new \Exception('Link sharing is disabled');
```
Comment (LukasReschke) [inline apps/files_sharing/lib/middleware/sharingcheckmiddleware.php]: Can we do a NotFoundException?

Comment (LukasReschke) [inline apps/files_sharing/lib/middleware/sharingcheckmiddleware.php]: Reason is https://github.com/owncloud/core/blob/d74568dab8b230601f6e63c5fb0c50ae444f218c/apps/files_sharing/lib/middleware/sharingcheckmiddleware.php#L89-L95

Comment (rullzer) [inline apps/files_sharing/lib/middleware/sharingcheckmiddleware.php]: fixed

**Comment by rullzer** [comment]
Ok the middleware now properly handles the externalshares stuff as well.

* For the externalshares the class uses 2 annotations. By default all methods require incomming and outgoing s2s shares to be enabled. Annotation only disable the check.
* Since it is an API endpoint do not show the generic error page, rather return a 405
* Fix and updated unit tests (coverage of the middleware at 100%)

Review time!

CC: @LukasReschke @MorrisJobke @PVince81 @nickvergessen @DeepDiver1975

**Comment by schiesbn** [comment]
great work, code looks good :+1:

**Comment by PVince81** [comment]
Makes sense :+1:

**Comment by MorrisJobke** [comment]
@rullzer Is it possible to backport this to stable8.1? https://github.com/owncloud/core/issues/19477#issuecomment-144657239

**Comment by rullzer** [comment]
@MorrisJobke how badly do we want to backport this? Since there were multiple changes for 8.2 in those files. I can backport them all in 1 (they are all pretty clean). But only this one is kind of messy.

**Comment by MorrisJobke** [comment]
> @MorrisJobke how badly do we want to backport this? Since there were multiple changes for 8.2 in those files. I can backport them all in 1 (they are all pretty clean). But only this one is kind of messy.

@michaelstingl Can you ellaborate on this?

**Comment by michaelstingl** [comment]
@MorrisJobke @rullzer A backport to 8.1.4 would be awesome! (Important for the user due to legal
reasons)

**Comment by rullzer** [comment]
backport in #19638

**Pull request merged**